### PR TITLE
Fix: Form reset custom field default value

### LINKED ISSUE #5297: form reset custom field with value per default
What type of report is this:

| Q  | A
| ---| ---
| Bug report? | X
| Feature request? | 
| Enhancement? | 

## Description:
Submitting a form reset custom field of a contact with value per default

## If a bug:

| Q   | A
| --- | ---
| Mautic version | 2.10
| PHP version | 

### Steps to reproduce:
1. Create custom field "text" type and fill a default value
2. Create a contact and fill this new custom field with a value (not the default one)
3. Create a form (but do not add this new custom field in the form). For example email and lastname only
4. Submit the form
5. Check the contact 
6. The new value of this custom field above has been updating with the default value
 
### Log errors: 

_Please check for related errors in the latest log file in [mautic root]/app/log/ and/or the web server's logs and post them here. Be sure to remove sensitive information if applicable._ 

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/51256481-form-reset-custom-field-with-value-per-default?utm_campaign=plugin&utm_content=tracker%2F5355074&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F5355074&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by kuzmany**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | y
| New feature? | 
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | fixes https://github.com/mautic/mautic/issues/5297
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Notice issue with forms  because on each submit we call `$this->leadModel->getCurrentLead()`  what mean, If tracking cookie doesn't exist, then create new lead. That means, new created contact  with default values is merged to old one, and rewrite options with default values.

My solution, after form submit If contact already exist and tracking cookie doesn't exist,  then we try merge created contact with exist contact (https://github.com/mautic/mautic/pull/5380/files#diff-93de7d27865690af0cae5ea03e7b9433R998). But  if we use default  value for contact fields, then this default values reset value on existed contact. 

[//]: # ( As applicable: )
#### Steps to reproduce the bug:

1. Create a custom field for contacts with a custom default value
2. Create a contact in Mautic including an email address (using quick add or regular add) and modify the custom field (change the default value)
3. Create a form without the custom field being on the form (just the email field for example) and submit the form with the same email address as your contact created in step 2
4. Notice that the custom field value was changed to the default value even though you edited it in step 2, because the  form submission creates a contact, then tries to find existing contacts with unique field (email) matching and then merges two profiles, overwriting the value that was set.

This bugfix/feature has been sponsored by @Webmecanik

#### Steps to test this PR:
1.  Repeat all steps
2.  Custom field value of existing contact should not be rewritten with the default value. It should still keep the original value (whether empty or not)

### CODE REVIEW

**Comment by npracht** [comment]
Tested it solved this bug !

**Comment by alanhartless** [comment]
Just from reading the code, I think this will break manually merging contacts as any field with a default set will be ignored. (Go to a contact's details page and click the Merge button).

**Comment by kuzmany** [comment]
@alanhartless  I did changes. Now It's more clear. I've added new param $onlyFields to mergeLeads  and if lead entity is new then mautic will merge just passed params throught $onlyFields array.

**Comment by kuzmany** [comment]
Label: Ready to test

**Comment by npracht** [comment]
@justinfortes it has been modified as requested by Alan. Could we add that as candidate for 2.13.0 keeping in mind that it is a bugfix !
Thanks !

File: app/bundles/LeadBundle/Model/LeadModel.php
Comment (dbhurley) [inline app/bundles/LeadBundle/Model/LeadModel.php]: Just at a quick read through, there's only one foreach here vs 2 in the original. Was this intended?

Comment (kuzmany) [reply app/bundles/LeadBundle/Model/LeadModel.php]: It's ok.
Before fields were  loaded by   `$mergeFromFields = $mergeFrom->getFields();`
After PR  `$mergeFromFields = $this->leadFieldModel->getFieldListWithProperties($mergeFrom);`
Each function return different arrays, I replaced it, it's better readable.

Comment (dbhurley) [reply app/bundles/LeadBundle/Model/LeadModel.php]: Okay, long as nothing was missed. Thanks!

**Review by npracht — DISMISSED** [review]
We retested it after the several changes, the fix is still working as it was before and we couldn't generate an issue coming from that.

**Comment by npracht** [comment]
Label: Pending test confirmation

**Comment by npracht** [comment]
Label: Code Review

**Comment by javjim-mautic** [comment]
I don't know if I did something wrong but I couldnt reproduce the bug. It worked ok before the PR.

And after PR it errors out

@kuzmany

**Comment by kuzmany** [comment]
Label: WIP

**Comment by kuzmany** [comment]
Label: Pending Feedback

**Comment by kuzmany** [comment]
@javjim-mautic I will retest againg, If isssue exist, I will record screencast

**Comment by kuzmany** [comment]
@javjim-mautic  Just tested on actual staging and can't reproduce too. Looks like another pull request resolve that  issue.
@npracht  Can you test it?

**Comment by npracht** [comment]
Hello @kuzmany & @javjim-mautic !
I'm not able to reproduce the issue anymore (before PR). I agree with Zdeno, since this is a PR dating from last november, it may have been corrected by another PR?...

**Comment by kuzmany** [comment]
Then I'll close it

**Comment by johbuch** [comment]
@kuzmany 
can you try this one again ?
I just tried again and default value erased other values.
thanks

**Comment by kuzmany** [comment]
@johbuch confirmed.
Try now test.

**Comment by johbuch** [comment]
@kuzmany just tested and it works. Good job ;)

**Comment by escopecz** [comment]
@kuzmany could you merge/rebase?

File: app/bundles/FormBundle/Tests/FormTestAbstract.php
Comment (escopecz) [inline app/bundles/FormBundle/Tests/FormTestAbstract.php]: Class Mautic\FormBundle\Tests\ContactTracker not found.

**Review by alanhartless — CHANGES_REQUESTED** [review]
Behavior prior to this change was that conflicting unique identifiers would result in a new contact. Now if you submit the same form several times in a row with different email addresses, it updates the tracked contact regardless. I think we should keep the existing behavior of error-ing on the side of caution and not merge in the case of conflicting unique identifiers.

**Comment by johbuch** [comment]
@kuzmany have you seen the previous comment from alan ?
I tested the kiosk mode with form and it does not work anymore due to this PR.
Would it be possible to make a fix of this PR so that it cannot create another bug with form and kiosk mode ?
thanks

**Comment by kuzmany** [comment]
@Maxell92 can you get code review for ContactTracker https://github.com/mautic/mautic/pull/5380/files#diff-b3bd4fab79438ede27cbd2496324e534R115 (assume should return null if not device exist?)

**Comment by kuzmany** [comment]
Reworked (kiosk mode should works either)

@johbuch please test
@alanhartless I made changes. 
Disable kiosk: 
Now we try find contact by tracking ID, if not exist we create new Lead entity. Then in next steps If we found lead by unique fields, then we will merge lead to foundedLead. 

Before this change leadModel->getCurrentLead() - create new Lead with setEntityDefaultValues.

Please review it.

**Review by npracht — DISMISSED** [review]
Just tested, kiosk mode works and the reset issue is fixed.
Good job !

**Review by npracht — CHANGES_REQUESTED** [review]
@kuzmany it does not work properly 🔴 
If you submit in kiosk mode the form with the same email, it creates a duplicate.
If the email already exists, it should update the existing contact with the same email address.

Capture d’écran 2019-03-26 à 09 44 06
Capture d’écran 2019-03-26 à 09 44 19

**Comment by kuzmany** [comment]
@npracht but this is how  Kiosk  work, isn't it?

> For example, having contacts submit forms at a trade show or conference. When the kiosk mode is turned on, each submission will create a new contact

**Comment by npracht** [comment]
Yes, it does not take into account the tracking information. But, it does not create duplicates.
The objective is that if an ipad is given for salesteam with a mautic form, they can submit new contact information. That way with kiosk mode enabled, all the contact are not merged (even with different email address, what would happen without kiosk mode). But if a contact already exist (same email address) it should update the existing contact.

**If i make it short, kiosk mode is only expected to do not merge by tracking ID.**

**Comment by kuzmany** [comment]
@npracht OK, that is mine misunderstanding Kiosk mode from documentation.
Last commit should fixed it.

**Review by npracht — DISMISSED** [review]
Finally working properly.

**Comment by npracht** [comment]
@kuzmany can you check what's wrong with Travis ?

**Comment by kuzmany** [comment]
@npracht wait for feedback to this change https://github.com/mautic/mautic/pull/5380/files#diff-b3bd4fab79438ede27cbd2496324e534R115
Maybe I didn't catch context, and we will change it.
Then I will modify unit tests and travis check

**Comment by escopecz** [comment]
@kuzmany Travis CI is reporting that there are some tests failing.

**Comment by kuzmany** [comment]
@escopecz I know. Because I change this line, I think that condition is wrong https://github.com/mautic/mautic/pull/5380/files#diff-b3bd4fab79438ede27cbd2496324e534R115 

I am trying getting feedback from your team about that before I spent more times on unit tests. Can you check it?

**Comment by escopecz** [comment]
I checked with @alanhartless:

> I'm pretty sure that's the transition code that closed the vulnerability to spoof tracking of someone. The idea was that the first time was allowed but after that, i.e., once they had a tracked device, we no longer recognized them by the mtc_id because that re-introduces the vulnerability.
I'm pretty sure by changing that line, we re-introduce the vulnerability to allow tracking by setting the cookie mtc_id to the contact's ID which can be incremented by 1 to find all contacts.
Before we introduced that code, no contact had a tracked device. That's why it's checking if they have a device and if so, we don't honor the mtc_id because it could be spoofed.

**Comment by kuzmany** [comment]
@npracht @johbuch can you re-test? 
We had to make changes prior to this comment https://github.com/mautic/mautic/pull/5380#issuecomment-486415145

**Comment by npracht** [comment]
Hi there!
It has been decided to not create any extra Mautic 2 versions (except for major security purpose) knowing that Mautic 3 is coming very soon.

**We now want to integrate your contribution in the Mautic 3 roadmap as 3.0.1 candidate.**

## How to do?
1. Check if the bug still present in Mautic 3
2. **If Yes**: rebase your PR against the 3.x branch
3. **If No**: please comment on your PR: “This PR only applies to Mautic 2.x”

**Please report results by commenting on your PR to make us administration easier.**

_In case your bugfix only apply to Mautic 2, we'll consider adding it in an extra Mautic 2 version._
______
You can more information on how to do all of that on [this blog post "Getting you PR ready for Mautic 3"](https://www.mautic.org/blog/developer/getting-your-pr-ready-mautic-3).

**Comment by npracht** [comment]
@kuzmany can you look at conflicts please ?

**Comment by kuzmany** [comment]
This issue still exists on M3. 
I rebased it

**Comment by RCheesley** [comment]
As this PR [decreases coverage](https://github.com/mautic/mautic/pull/5380#issuecomment-650741057) we need to have tests such that it either does not change, or ideally improves coverage. Bumping to 3.1 as a result.

**Comment by RCheesley** [comment]
@cla-bot check

**Comment by RCheesley** [comment]
Bumping to 3.1.2 due to test coverage needing to be improved.

**Comment by npracht** [comment]
@kuzmany please increase code coverage.

**Comment by npracht** [comment]
@kuzmany any chance to add a test here and get rid of this PR :)

**Comment by kuzmany** [comment]
This is not issue anymore. 
Fixed by https://github.com/mautic/mautic/pull/7717

**Pull request closed without merging**
